git/client.go

```diff
@@ -0,0 +1,581 @@
+package git
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/url"
+	"os/exec"
+	"path"
+	"regexp"
+	"runtime"
+	"sort"
+	"strings"
+
+	"github.com/cli/cli/v2/internal/run"
+	"github.com/cli/safeexec"
+)
+
+var remoteRE = regexp.MustCompile(`(.+)\s+(.+)\s+\((push|fetch)\)`)
+
+// ErrNotOnAnyBranch indicates that the user is in detached HEAD state.
+var ErrNotOnAnyBranch = errors.New("git: not on any branch")
+
+type NotInstalled struct {
+	message string
+	error
+}
+
+func (e *NotInstalled) Error() string {
+	return e.message
+}
+
+type gitCommand struct {
+	*exec.Cmd
+}
+
+// This is a hack in order to not break the hundreds of
+// existing tests that rely on `run.PrepareCmd` to be invoked.
+func (gc *gitCommand) Run() error {
+	return run.PrepareCmd(gc.Cmd).Run()
+}
+
+// This is a hack in order to not break the hundreds of
+// existing tests that rely on `run.PrepareCmd` to be invoked.
+func (gc *gitCommand) Output() ([]byte, error) {
+	return run.PrepareCmd(gc.Cmd).Output()
+}
+
+type Client struct {
+	GhPath    string
+	RepoDir   string
+	GitPath   string
+	AuthHosts []string
+	Stderr    io.Writer
+	Stdin     io.Reader
+	Stdout    io.Writer
+
+	commandContext func(ctx context.Context, name string, args ...string) *exec.Cmd
+}
+
+func (c *Client) Command(args ...string) (*gitCommand, error) {
+	if c.RepoDir != "" {
+		args = append([]string{"-C", c.RepoDir}, args...)
+	}
+	if c.commandContext == nil {
+		c.commandContext = exec.CommandContext
+	}
+	if c.GitPath == "" {
+		var err error
+		c.GitPath, err = safeexec.LookPath("git")
+		if err != nil {
+			if errors.Is(err, exec.ErrNotFound) {
+				programName := "git"
+				if runtime.GOOS == "windows" {
+					programName = "Git for Windows"
+				}
+				return nil, &NotInstalled{
+					message: fmt.Sprintf("unable to find git executable in PATH; please install %s before retrying", programName),
+					error:   err,
+				}
+			}
+			return nil, err
+		}
+	}
+	cmd := c.commandContext(context.Background(), c.GitPath, args...)
+	cmd.Stderr = c.Stderr
+	cmd.Stdin = c.Stdin
+	cmd.Stdout = c.Stdout
+	return &gitCommand{cmd}, nil
+}
+
+// AuthenticatedCommand is a wrapper around Command that included configuration to use gh
+// as the credential helper for git.
+func (c *Client) AuthenticatedCommand(args ...string) (*gitCommand, error) {
```

Comment:
I did this mostly to not deal with hundreds of tests that check args for git invocations. I would have had to go and insert `-c credential.helper...` for each of these tests. I agree, in the future it would be nice to just include these in all invocations.